cmd/namespace/validate.go

```diff
@@ -43,44 +64,145 @@ var configSchema *jsonschema.Schema
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Encountered unmarshal error for \"%s\": %+v\n", name, err)
 		return nil, cmdx.FailSilently(cmd)
 	}
 
 	if err := configSchema.ValidateInterface(val); err != nil {
+		fmt.Fprintf(cmd.ErrOrStderr(), "File %s was not a valid namespace file. Reasons:\n", name)
 		jsonschemax.FormatValidationErrorForCLI(cmd.ErrOrStderr(), config.Schema, err)
 		return nil, cmdx.FailSilently(cmd)
 	}
 
 	var n namespace.Namespace
-	if err := yaml.Unmarshal(fc, &n); err != nil {
-		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Encountered yaml unmarshal error for \"%s\": %+v\n", fn, err)
+	if err := yaml.Unmarshal(b, &n); err != nil {
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Encountered unmarshal error for \"%s\": %+v\n", name, err)
 		return nil, cmdx.FailSilently(cmd)
 	}
 
 	return &n, nil
 }
+
+func validateConfigFile(cmd *cobra.Command, fn string) error {
+	fc, err := ioutil.ReadFile(fn)
+	if err != nil {
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Could not read file \"%s\": %+v\n", fn, err)
+		return cmdx.FailSilently(cmd)
+	}
+
+	// TODO: handle other config formats (json, toml, etc)
+	var val map[string]interface{}
+	dot := strings.LastIndex(fn, ".")
+	if dot == -1 {
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Unable to infer file type from \"%s\": %+v\n", fn, err)
+		return cmdx.FailSilently(cmd)
+	}
+
+	var unmarshal func(b []byte, v interface{}) error
+	ext := fn[dot+1:]
+	switch ext {
+	case "yaml", "yml":
+		unmarshal = yaml.Unmarshal
+	case "json":
+		unmarshal = json.Unmarshal
+	case "toml":
+		unmarshal = toml.Unmarshal
+	default:
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Unhandled config file extension \"%s\"\n", ext)
+		return cmdx.FailSilently(cmd)
+	}
+
+	if err := unmarshal(fc, &val); err != nil {
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Encountered unmarshal error for \"%s\": %+v\n", fn, err)
+		return cmdx.FailSilently(cmd)
+	}
+
+	if ns, ok := val["namespaces"]; ok {
+		// ns can either be a ws url or a list of namespace objects)
+		switch t := ns.(type) {
+		case string:
+			logger := logrusx.New("cmd", "0")
+			cw, err := config.NewNamespaceWatcher(context.Background(), logger, ns.(string))
+			if err != nil {
+				_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Encountered error reading config: %+v\n", err)
+				return cmdx.FailSilently(cmd)
+			}
+
+			files := cw.NamespaceFiles()
+			for _, file := range files {
+				_, err := validateNamespaceBytes(cmd, file.Name, file.Contents, file.Parser)
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		case []interface{}:
+			for i, obj := range ns.([]interface{}) {
+				fc, err := yaml.Marshal(obj)
+				if err != nil {
+					_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Internal error. Failed to marshal yaml %+v\n", err)
+					return cmdx.FailSilently(cmd)
+				}
+				_, err = validateNamespaceBytes(cmd, fmt.Sprintf("index: %d", i), fc, yaml.Unmarshal)
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		case []map[string]interface{}:
```

Comment:
note: exact same code as above, it just didn't feel right to try to generalize where a simple copy paste seemed to be readable enough. (toml parser returns `[]map[string]interface{}` instead of `[]interface{}`. Interestingly Go does not let you cast `[]map[string]interface{}` as `[]interface{}` which breaks my assumption that `interface{}` stands for anything including `map`)

Reply:
That makes sense. I was thinking that `[]interface{}` expressed "an array of anything" like in Java `List<?>`, but that's not how the compiler sees it